resources/bundles/org.eclipse.core.filesystem/src/org/eclipse/core/internal/filesystem/InternalFileSystemCore.java

```diff
@@ -72,9 +81,15 @@ public IFileSystem getFileSystem(String scheme) throws CoreException {
 			IConfigurationElement element = (IConfigurationElement) result;
 			FileSystem fs = (FileSystem) element.createExecutableExtension("run"); //$NON-NLS-1$
 			fs.initialize(scheme);
-			//store the file system instance so we don't have to keep recreating it
-			registry.put(scheme, fs);
-			return fs;
+			synchronized (this) {
+				result = registry.get(scheme);
```

Comment:
nit: could use computeIfAbsent instead of get/put, but not important

Reply:
The bad thing is that map **can** contain value but if it is not a FileSystem, we still have to put here.